src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts

```diff
@@ -55,47 +66,72 @@ export class LocalPythonAndRelatedNonPythonKernelSpecFinder extends LocalKernelS
         @inject(IMemento) @named(GLOBAL_MEMENTO) globalState: Memento
     ) {
         super(fs, workspaceService, extensionChecker, globalState);
+        interpreterService.onDidChangeInterpreters(() => this.refresh().catch(noop), this, this.disposables);
+        kernelSpecsFromKnownLocations.onDidChangeKernels(() => this.refresh().catch(noop), this, this.disposables);
+        interpreterService.onDidChangeInterpreter(() => this.refresh().catch(noop), this, this.disposables);
     }
-    public async listKernelSpecs(resource: Resource, ignoreCache?: boolean, cancelToken?: CancellationToken) {
```

Comment:
No more pull, its all push based, we have an event and a property to access the kernels.